fix: ux behavior on save for anonymous users

nikochiko opened the pull request:
should redirect to login page with next=...

### Q/A checklist

- [x] If you add new dependencies, did you update the lock file?
```bash
poetry lock --no-update
```
- [x] Run tests 
```bash
ulimit -n unlimited && ./scripts/run-tests.sh
```
- [x] Do a self code review of the changes - Read the diff at least twice.  
- [x] Carefully think about the stuff that might break because of this change - this sounds obvious but it's easy to forget to do "Go to references" on each function you're changing and see if it's used in a way you didn't expect. 
- [x] The relevant pages still run when you press submit
- [x] The API for those pages still work (API tab)
- [x] The public API interface doesn't change if you didn't want it to (check API tab > docs page)
- [x] Do your UI changes (if applicable) look acceptable on mobile?
- [x] Ensure you have not regressed the import time unless you have a good reason to do so. 
You can visualize this using tuna:
```bash
python3 -X importtime -c 'import server' 2> out.log && tuna out.log
```
To measure import time for a specific library:
```bash
$ time python -c 'import pandas'

________________________________________________________
Executed in    1.15 secs    fish           external
   usr time    2.22 secs   86.00 micros    2.22 secs
   sys time    0.72 secs  613.00 micros    0.72 secs
```
To reduce import times, import libraries that take a long time inside the functions that use them instead of at the top of the file:
```python
def my_function():
    import pandas as pd
    ...
```

### Legal Boilerplate

Look, I get it. The entity doing business as “Gooey.AI” and/or “Dara.network” was incorporated in the State of Delaware in 2020 as Dara Network Inc. and is gonna need some rights from me in order to utilize my contributions in this PR. So here's the deal: I retain all rights, title and interest in and to my contributions, and by keeping this boilerplate intact I confirm that Dara Network Inc can use, modify, copy, and redistribute my contributions, under its choice of terms.



---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1208189770838403

devxpy commented:
curious if this is upto date as per 
> @SeanBlagsvedt This is the intended UX flow for anonymous users.

devxpy commented on daras_ai_v2/base.py:
```diff
@@ -514,23 +520,35 @@ def _render_published_run_save_buttons(
-                        is_update_mode=can_edit,
                     )
 
+    def _save_for_anonymous_user(self, current_run: SavedRun):
```
believe you dont need to pass current_run around anymore?

devxpy commented on daras_ai_v2/base.py:
```diff
@@ -514,23 +520,35 @@ def _render_published_run_save_buttons(
                     )
 
+    def _save_for_anonymous_user(self, current_run: SavedRun):
+        if self._has_request_changed():
```
I believe what @SeanBlagsvedt asked for is that the run not be actually charged under anon but under the logged in user?

devxpy commented on daras_ai_v2/base.py:
```diff
@@ -514,23 +520,35 @@ def _render_published_run_save_buttons(
+            if not current_run:
+                gui.rerun()
+
+        self.ensure_authentication(
```
this should do only redirect not the check for auth given this function is called `_save_for_anonymous_user`

devxpy commented on daras_ai_v2/base.py:
```diff
@@ -607,46 +624,75 @@ def _render_publish_modal(
-        if not pressed_save:
-            return
+        update_existing_published_run = self.can_user_edit_published_run(published_run)
+        if pressed_save:
```
this if statement is better written as `if not <condn>: return`

devxpy commented on daras_ai_v2/base.py:
```diff
@@ -607,46 +624,75 @@ def _render_publish_modal(
 
-        if not pressed_save:
-            return
+        update_existing_published_run = self.can_user_edit_published_run(published_run)
```
only need to check this after pressed_save condition

devxpy commented on daras_ai_v2/base.py:
```diff
@@ -607,46 +624,75 @@ def _render_publish_modal(
-                visibility=published_run_visibility,
+                title=title or published_run.title,
+                notes=(
+                    notes if notes is not None else published_run.notes
```
inline if condition discouraged

devxpy commented on daras_ai_v2/base.py:
```diff
@@ -607,46 +624,75 @@ def _render_publish_modal(
+                        gui.session_state["published_run_title"]
+                    )
+                except TitleValidationError as e:
+                    gui.error(str(e))
+                    return
 
-        is_root_published_run = is_update_mode and published_run.is_root()
-        if not is_root_published_run:
-            try:
-                self._validate_published_run_title(published_run_title)
-            except TitleValidationError as e:
-                gui.error(str(e))
-                return
+            self.save_or_update_published_run_and_redirect(
+                current_run=current_run,
+                published_run=(
+                    published_run if update_existing_published_run else None
+                ),
+                title=published_run_title.strip(),
+                notes=published_run_notes.strip(),
+                visibility=published_run_visibility,
+            )
```
refactor as `if ... else`

devxpy commented on daras_ai_v2/base.py:
```diff
@@ -607,46 +624,75 @@ def _render_publish_modal(
+                # don't validate title root example
+                try:
+                    self._validate_published_run_title(
+                        gui.session_state["published_run_title"]
```
use local var `published_run_title`

nikochiko commented:
> curious if this is upto date as per
> 
> > @SeanBlagsvedt This is the intended UX flow for anonymous users. https://docs.google.com/presentation/d/1P5MKF0jykYPw-G_9SgvKEpu7fjD2v-0gZgLJB4Uw174/edit#slide=id.g2ef5e525406_0_329

needs an update, if I remember right

devxpy commented on daras_ai_v2/base.py:
```diff
@@ -343,6 +345,12 @@ def render(self):
         self._user_disabled_check()
         self._check_if_flagged()
 
+        if self.should_submit_after_login():
```
SAVE_AFTER_LOGIN_Q and SUBMIT_AFTER_LOGIN_Q shold be mutually exclusive

nikochiko replied:
done

devxpy commented on daras_ai_v2/base.py:
```diff
@@ -99,6 +87,8 @@
 gooey_rng = Random()
 
 SUBMIT_AFTER_LOGIN_Q = "submitafterlogin"
+SAVE_AFTER_LOGIN_Q = "saveafterlogin"
```
The save terminology is very confusing given we are using published run everywhere else. Use `PUBLISH_AFTER_LOGIN_Q`

devxpy commented on daras_ai_v2/base.py:
```diff
@@ -1621,6 +1723,13 @@ def should_submit_after_login(self) -> bool:
             and not self.request.user.is_anonymous
         )
 
+    def should_save_after_login(self) -> bool:
```
should_create_published_run_after_login

devxpy commented on daras_ai_v2/base.py:
```diff
@@ -620,28 +705,36 @@ def _render_publish_form(
 
+    def _get_default_pr_title(self):
+        recipe_title = self.get_root_pr().title or self.title
+        if self.request.user and not self.request.user.is_anonymous:
```
I believe this should just not run at all if the user is null/anon?

devxpy commented on daras_ai_v2/base.py:
```diff
@@ -1594,11 +1672,35 @@ def render_extra_waiting_output(self):
+        else:
+            raise gui.RedirectException(self.app_url(run_id=sr.run_id, uid=sr.uid))
+
+    def save_and_redirect(self, sr: SavedRun | None = None) -> typing.NoReturn:
```
create_published_run_and_redirect

nikochiko commented on daras_ai_v2/base.py:
```diff
@@ -325,6 +315,17 @@ def sentry_event_set_user(self, event, hint):
+            return
+
+        try:
+            request_state = base64.urlsafe_b64decode(request_state).decode()
```
but for anon users, we won't have their real `uid` until they login.

are you suggesting that we create a run without a `uid`, and when the user comes back with `SUBMIT_AFTER_LOGIN_Q`, we then update the `uid` and run?

devxpy replied:
we generate a firebase user as soon as you submit, this should do the same thing

nikochiko replied:
for anon users who come back with an existing account, we would have 2 different uids.

i fixed this case by setting the uid of the unqueued run when user comes back. i added a concurrency check there too so that we don't queue the same run twice.

nikochiko commented on daras_ai_v2/base.py:
```diff
@@ -1594,11 +1672,35 @@ def render_extra_waiting_output(self):
+            published_run_id=get_random_doc_id(),
+            saved_run=sr or self.current_sr,
+            user=self.request.user,
+            title=gui.session_state.get(
```
fixed

devxpy commented on daras_ai_v2/base.py:
```diff
@@ -490,39 +574,49 @@ def _render_published_run_save_buttons(self, *, sr: SavedRun, pr: PublishedRun):
                 className="mb-0 ms-lg-2 px-lg-4",
                 type="primary",
             ):
+                if not self.request.user or self.request.user.is_anonymous:
```
please put this in an if..else case |  its not clear from this that publish_for_anonymous_user will end up raising an exception and the stuff below won't run

nikochiko replied:
i added a type annotation for `_publish_for_anonymous_user() -> NoReturn` for this, but i like this because it's even more explicit

devxpy commented on daras_ai_v2/base.py:
```diff
@@ -1621,10 +1706,42 @@ def should_submit_after_login(self) -> bool:
+        self,
+        *,
+        enable_rate_limits: bool = False,
+        run_status: str | None = "Starting...",
```
not used

devxpy commented on daras_ai_v2/base.py:
```diff
@@ -490,39 +574,49 @@ def _render_published_run_save_buttons(self, *, sr: SavedRun, pr: PublishedRun):
+    def _publish_for_anonymous_user(self):
+        query_params = {PUBLISH_AFTER_LOGIN_Q: "1"}
+        if self._has_request_changed():
+            sr = self.create_new_run(enable_rate_limits=True, run_status=None)
```
did you mean create_and_validate_run?

nikochiko replied:
ah yes, thanks. good catch

devxpy commented:
is this finally on me? I saw 2 commits being pushed after it was assigned to me

nikochiko commented:
> is this finally on me? I saw 3 commits being pushed after it was assigned to me

later two were fixes from the review

devxpy commented:
> > is this finally on me? I saw 3 commits being pushed after it was assigned to me
> 
> yes, later two were fixes from the review

sorry i didn't realise the review got out

devxpy commented:
odd behavior
- click on share
- change visibility
- press copy link
- close dialog
- open dialog again, see that visibility hasn't saved

devxpy commented:
<img width="575" alt="image" src="https://github.com/user-attachments/assets/e92d8cfa-6ee6-467d-bb25-940349ccea93">

close button should be top right corner aligned like so 

<img width="1178" alt="image" src="https://github.com/user-attachments/assets/d02fd264-9135-4944-8564-cc2a495a61dc">

devxpy commented:
dialog buttons should be aligned at the corners with equal spacing on bottom and left/right

<img width="833" alt="image" src="https://github.com/user-attachments/assets/2d209e76-ec74-41f3-96fd-de2c6a7e4c5f">
<img width="553" alt="image" src="https://github.com/user-attachments/assets/a08f960d-febe-4233-8d4e-12344dfdd4e6">

example of proper alignment:

<img width="532" alt="image" src="https://github.com/user-attachments/assets/f29bbe7b-e0ee-4c84-a69e-b5cd6705c4b2">

devxpy commented:
inconsistency of display between update and options buttons:

- update button shown on all tabs
- options button shown on run, run as API and integrations only

devxpy commented on daras_ai_v2/base.py:
```diff
@@ -1164,7 +1217,7 @@ def get_sr_pr_from_query_params(
         cls,
         run_id: str,
-        uid: str,
+        uid: str | None,
```
not sure why this annotation has changed, it shouldn't be

nikochiko commented:
> odd behavior
> 
> * click on share
> * change visibility
> * press copy link
> * close dialog
> * open dialog again, see that visibility hasn't saved

That sounds like a reasonable user flow. I'll make it such that the `Copy Link` button also saves like the other `Save` button?

nikochiko commented:
<img width="697" alt="image" src="https://github.com/user-attachments/assets/d645cb72-55dc-4406-8730-0da6d16b438b">

we have bad defaults for padding/margin of headings, this looks uneven too (without any changes made). i'll fix it by adding some top margin/padding to the heading's container too

nikochiko commented:
better?

<img width="925" alt="image" src="https://github.com/user-attachments/assets/98525493-4fa2-4634-af4c-020fad7313ee">

<img width="925" alt="image" src="https://github.com/user-attachments/assets/cd8ec6a5-c060-46d6-a66b-1edde5e6b2b0">

nikochiko commented on daras_ai_v2/base.py:
```diff
@@ -1594,23 +1647,57 @@ def render_extra_waiting_output(self):
+        assert self.request.user and not self.request.user.is_anonymous
+
+        if (
+            self.get_run_state(gui.session_state) == RecipeRunState.starting
```
nice, i like this refactor.

yes, `RecipeRunState.starting` was misleading. I initially created a new state `not_started` and later realised that we had `starting` for that already.

devxpy commented on daras_ai_v2/base.py:
```diff
@@ -1594,23 +1647,57 @@ def render_extra_waiting_output(self):
+                    # run belonged to an anonymous user who now logged
+                    # in from an existing account now, so we update the uid
+                    sr.uid = self.request.user.uid
+                    sr.save(update_fields=["uid"])
```
PS didn't need to save here as `dump_state_to_sr` would have done it  anyways

nikochiko replied:
i'm aware, but i added this because it isn't obvious (also would not be obvious to fix if we were to later refactor `dump_state_to_sr` to only save the state and not everything else)

devxpy commented on daras_ai_v2/base.py:
```diff
@@ -1594,23 +1647,57 @@ def render_extra_waiting_output(self):
+    def publish_and_redirect(self) -> typing.NoReturn | None:
+        assert self.request.user and not self.request.user.is_anonymous
+
+        if (
```
This transaction stuff is more complex than it needs to be. You can do something like:

```py
matches = SavedRun.objects.filter(
    id=CURRENT_RUN.ID,
    run_status="", error_msg="", run_time=datetime.timedelta()  # filter for pending state
).update(run_status=STARTING_STATE, uid=LOGGED_IN_USER.UID)
if matches > 0:  # no matches, run has probably started elsewhere
    >> send task to queue, create published run & redirect
else:
    >> rerun
```

nikochiko replied:
thanks

nikochiko commented:
> That sounds like a plausible user flow. I'll make it such that the Copy Link button also saves like the other Save button

this would require more work than I had thought. right now it doesn't trigger a form submit - all javascript.

nikochiko commented:
fixed the rest, other than the Copy Link fix.

i'll need to release gooey-gui and update version for that.

devxpy commented:
> > That sounds like a plausible user flow. I'll make it such that the Copy Link button also saves like the other Save button
> 
> this would require more work than I had thought. right now it doesn't trigger a form submit - all javascript.

You can execute js on a python button I believe
```
if button_pressed:
   gui.js(<js code here>)
```

devxpy commented on daras_ai_v2/copy_to_clipboard_button_widget.py:
```diff
@@ -23,20 +22,25 @@ def copy_to_clipboard_button(
-        """,
+    key: str | None = None,
+) -> bool:
+    key = key or "copy-to-clipboard-btn"
```
lets not make the other copy buttons slower

nikochiko replied:
fixed

nikochiko replied:
added a new method `copy_to_clipboard_button_with_return` (had to use chatgpt for name suggestions)

devxpy commented on daras_ai_v2/base.py:
```diff
@@ -1594,23 +1654,46 @@ def render_extra_waiting_output(self):
+
+        updated_count = SavedRun.objects.filter(
+            id=self.current_sr.id,
+            uid=self.current_sr_user.uid,
```
this filter wont work if you login with a different user?

nikochiko replied:
`current_sr_user` is different from the current user (it's taken from `current_sr.uid`)

nikochiko replied:
`current_sr.uid` would be clearer here though, thanks!

devxpy replied:
why does this need the uid filter at all? since its filtering on `id`

devxpy commented:
I get this error when I try to press the options button as a fresh anonymous user 

```
Traceback (most recent call last):
  File "/Users/dev/.virtualenvs/ddgai-b0692ca2/lib/python3.10/site-packages/uvicorn/protocols/http/httptools_impl.py", line 404, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
  File "/Users/dev/.virtualenvs/ddgai-b0692ca2/lib/python3.10/site-packages/uvicorn/middleware/proxy_headers.py", line 78, in __call__
    return await self.app(scope, receive, send)
  File "/Users/dev/.virtualenvs/ddgai-b0692ca2/lib/python3.10/site-packages/fastapi/applications.py", line 270, in __call__
    await super().__call__(scope, receive, send)
  File "/Users/dev/.virtualenvs/ddgai-b0692ca2/lib/python3.10/site-packages/starlette/applications.py", line 124, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/Users/dev/.virtualenvs/ddgai-b0692ca2/lib/python3.10/site-packages/starlette/middleware/errors.py", line 184, in __call__
    raise exc
  File "/Users/dev/.virtualenvs/ddgai-b0692ca2/lib/python3.10/site-packages/starlette/middleware/errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "/Users/dev/Projects/dara/ddgai/./server.py", line 107, in middleware
    await app(scope, receive, send)
  File "/Users/dev/.virtualenvs/ddgai-b0692ca2/lib/python3.10/site-packages/starlette/middleware/sessions.py", line 86, in __call__
    await self.app(scope, receive, send_wrapper)
  File "/Users/dev/.virtualenvs/ddgai-b0692ca2/lib/python3.10/site-packages/starlette/middleware/authentication.py", line 48, in __call__
    await self.app(scope, receive, send)
  File "/Users/dev/.virtualenvs/ddgai-b0692ca2/lib/python3.10/site-packages/starlette/middleware/cors.py", line 92, in __call__
    await self.simple_response(scope, receive, send, request_headers=headers)
  File "/Users/dev/.virtualenvs/ddgai-b0692ca2/lib/python3.10/site-packages/starlette/middleware/cors.py", line 147, in simple_response
    await self.app(scope, receive, send)
  File "/Users/dev/.virtualenvs/ddgai-b0692ca2/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 75, in __call__
    raise exc
  File "/Users/dev/.virtualenvs/ddgai-b0692ca2/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 64, in __call__
    await self.app(scope, receive, sender)
  File "/Users/dev/.virtualenvs/ddgai-b0692ca2/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 21, in __call__
    raise e
  File "/Users/dev/.virtualenvs/ddgai-b0692ca2/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 18, in __call__
    await self.app(scope, receive, send)
  File "/Users/dev/.virtualenvs/ddgai-b0692ca2/lib/python3.10/site-packages/starlette/routing.py", line 680, in __call__
    await route.handle(scope, receive, send)
  File "/Users/dev/.virtualenvs/ddgai-b0692ca2/lib/python3.10/site-packages/starlette/routing.py", line 275, in handle
    await self.app(scope, receive, send)
  File "/Users/dev/.virtualenvs/ddgai-b0692ca2/lib/python3.10/site-packages/starlette/routing.py", line 65, in app
    response = await func(request)
  File "/Users/dev/.virtualenvs/ddgai-b0692ca2/lib/python3.10/site-packages/fastapi/routing.py", line 231, in app
    raw_response = await run_endpoint_function(
  File "/Users/dev/.virtualenvs/ddgai-b0692ca2/lib/python3.10/site-packages/fastapi/routing.py", line 162, in run_endpoint_function
    return await run_in_threadpool(dependant.call, **values)
  File "/Users/dev/.virtualenvs/ddgai-b0692ca2/lib/python3.10/site-packages/starlette/concurrency.py", line 41, in run_in_threadpool
    return await anyio.to_thread.run_sync(func, *args)
  File "/Users/dev/.virtualenvs/ddgai-b0692ca2/lib/python3.10/site-packages/anyio/to_thread.py", line 33, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
  File "/Users/dev/.virtualenvs/ddgai-b0692ca2/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 877, in run_sync_in_worker_thread
    return await future
  File "/Users/dev/.virtualenvs/ddgai-b0692ca2/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 807, in run
    result = context.run(func, *args)
  File "/Users/dev/Projects/dara/gooey-gui/py/gooey_gui/core/renderer.py", line 70, in wrapper
    return renderer(
  File "/Users/dev/Projects/dara/gooey-gui/py/gooey_gui/core/renderer.py", line 117, in renderer
    ret = render()
  File "/Users/dev/Projects/dara/ddgai/routers/root.py", line 615, in recipe_or_handle_or_static
    return render_recipe_page(request, parts[0], RecipeTabs.run, example_id)
  File "/Users/dev/Projects/dara/ddgai/routers/root.py", line 691, in render_recipe_page
    page.render()
  File "/Users/dev/Projects/dara/ddgai/./daras_ai_v2/base.py", line 362, in render
    self._render_header()
  File "/Users/dev/Projects/dara/ddgai/./daras_ai_v2/base.py", line 408, in _render_header
    self._render_options_button_with_dialog()
  File "/Users/dev/Projects/dara/ddgai/./daras_ai_v2/base.py", line 473, in _render_options_button_with_dialog
    self._unsaved_options_modal()
  File "/Users/dev/Projects/dara/ddgai/./daras_ai_v2/base.py", line 844, in _unsaved_options_modal
    assert self.request.user
AssertionError
```

devxpy commented on daras_ai_v2/base.py:
```diff
@@ -1594,23 +1661,46 @@ def render_extra_waiting_output(self):
+        ).update(run_status=STARTING_STATE, uid=self.request.user.uid)
+        if updated_count >= 1:
+            # updated now
+            self.dump_state_to_sr(self._get_validated_state(), self.current_sr)
```
not sure why this is needed here?

nikochiko replied:
ah, good catch. i think it is leftover from the earlier hack where the state was carried over without creating a saved run. thanks!

nikochiko commented:
@devxpy fixed both in https://github.com/GooeyAI/gooey-server/pull/481

devxpy commented:
I can still repro this in prod:

 - go to /copilot
 - make an edit to prompt
 - press save and run
 - login
 - run stuck on starting

nikochiko commented:
fixed in https://github.com/GooeyAI/gooey-server/pull/482 with `self.current_sr.refresh_from_db()` before queueing.

cause was that we updated `current_sr.uid` in the database with `SavedRun.objects.update()`, but not in the ORM object. we should probably get rid of `uid` from the runner task too. `SavedRun.id` should be good enough

devxpy commented:
this also broke rate limits fixed in https://github.com/GooeyAI/gooey-server/commit/7fe01cb4965cfe2dc5da33c7c015a13be6945231
